csaf/src/metadata_retriever/mod.rs

```diff
@@ -0,0 +1,200 @@
+            let host_url_string = format!("https://csaf.data.security.{}", host_url);
+            let dns_path = Url::parse(host_url_string.as_str())?;
+
+            match self.fetch_metadata_from_url(dns_path).await {
+                Ok(provider_metadata) => Ok(provider_metadata),
+                Err(err) => match err {
+                    Error::Fetcher(fetch_error) => match fetch_error {
+                        fetcher::Error::Request(reqwest_error) => {
+                            if reqwest_error.is_connect() {
+                                log::warn!("When DNS path URL {} fetch provider metadata fails due to a fetch error, which is a DNS error, the process will proceed to the next discovery rule. {:?}", provider_metadata_url.clone().to_string(), reqwest_error,);
+                                Ok(None)
+                            } else {
+                                Err(Error::Fetcher(fetcher::Error::Request(reqwest_error)))
+                            }
+                        }
+                    },
+                    _ => Err(err),
+                },
+            }
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+
+    /// Retrieve provider metadata through the security text of provided URL.
+    pub async fn retrieve_metadata_by_security_text(
+        &self,
+        provider_metadata_url: Url,
+        security_txt_path: &str,
+    ) -> Result<Option<Json<ProviderMetadata>>, Error> {
+        log::trace!(
+            "Starting retrieve provider metadata from security text of provided discovery URL "
+        );
+        if let Some(host_url) = provider_metadata_url.clone().host_str() {
+            let host_url_string = format!("https://{}", &host_url);
+            let host_url = Url::parse(host_url_string.as_str())?;
+            log::info!(
+                "Querying provider metadata url from security text of {:?}",
+                host_url_string
+            );
+            if let Some(path) = Self::get_metadata_url_from_security_text(
+                &self.fetcher,
+                host_url.join(security_txt_path)?,
+            )
+            .await?
+            {
+                let provider_metadata_url = Url::parse(path.as_str())?;
+
+                Ok(self.fetch_metadata_from_url(provider_metadata_url).await?)
+            } else {
+                log::warn!(
+                    "The Security Text obtained from this URL is 'None': {} ",
+                    host_url_string
+                );
+                Ok(None)
+            }
+        } else {
+            Err(Error::Url(ParseError::EmptyHost))
+        }
+    }
+
+    ///For the provided metadata URL, retrieve the provider metadata using three different discovery methods in a specified order.
+    pub async fn retrieve(&self) -> Result<DetectionType, Error> {
+        let actually_url = if self.base_url.clone().starts_with("https://") {
```

Comment:
I don't think it makes sense forcing this into a URL. Later on, it only calls .host_str() to extract it again.

Why not do:

```rust
if let Ok(url) = Url::parse(&self.base_url) {
  if url.schema != "https" { /*fail */ }
  // handle direct URL case
}
```

From what I see it, that would actually also make it unnecessary to pass that as an argument, but use `self.base_url` directly.

Reply:
I'm confused this one, it must be called .host_str(), but as we discussed, MetadataRetriever base_url type needs are string. so can we keep base_url is string in http.rs, then parse it to URL and change  MetadataRetriever base_url  is URL.   
`if let Ok(url) = Url::parse(&self.base_url) {
  if url.schema != "https" { /*fail */ }
  // handle direct URL case
}`. can I move the code into https?

Reply:
Why does it need to call `.host_str()`?

Reply:
https://github.com/ctron/csaf-walker/blob/10fed35a0ad25c1a1f0147bbc4d2b2071c67f263/csaf/src/metadata_retriever/mod.rs#L61
I need the base domain, then spell it into a complete URL. (I use host_str for get base domain)